core/src/main/java/org/mapstruct/SubClassMapping.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Configures the mapping to handle hierarchy of the source type.
+ * <p>
+ * The subclass to be mapped is to be specified via {@link #source()}.
+ * The subclass to map to is to be specified via {@link #target()}.
+ * </p>
+ * <p>
+ * This annotation can be combined with &#64;Mapping annotations.
+ * </p>
+ *
+ * <pre><code class='java'>
+ * &#64;Mapper
+ * public interface MyMapper {
+ *    &#64;SubClassMapping (sourceClass = SourceSubClass.class,
+ *                      targetClass = TargetSubClass.class)
+ *    TargetParent mapParent(SourceParent parent);
+ *
+ *    TargetSubClass mapSubClass(SourceSubClass subInstant);
+ * }
+ * </code></pre>
+ * Below follow examples of the implementation for the mapParent method.
+ * <strong>Example 1:</strong> For parents that cannot be created. (e.g. abstract classes or interfaces)
+ * <pre><code class='java'>
+ * // generates
+ * &#64;Override
+ * public TargetParent mapParent(SourceParent parent) {
+ *     if (parent instanceof SourceSubClass) {
+ *         return mapSubClass( (SourceSubClass) parent );
+ *     }
+ *     else {
+ *         throw new IllegalArgumentException("Not all subclasses are supported for this mapping. Missing for "
+ *                    + parent.getClass());
+ *     }
+ * }
+ * </code></pre>
+ * <strong>Example 2:</strong> For parents that can be created. (e.g. normal classes or interfaces with
+ * &#64;Mappper( uses = ObjectFactory.class ) )
+ * <pre><code class='java'>
+ * // generates
+ * &#64;Override
+ * public TargetParent mapParent(SourceParent parent) {
+ *     TargetParent targetParent1;
+ *     if (parent instanceof SourceSubClass) {
+ *         targetParent1 = mapSubClass( (SourceSubClass) parent );
+ *     }
+ *     else {
+ *         targetParent1 = new TargetParent();
+ *         // ...
+ *     }
+ * }
+ * </code></pre>
+ *
+ * @author Ben Zegveld
+ * @since 1.5
+ */
+@Repeatable( value = SubClassMappings.class )
+@Retention( RetentionPolicy.CLASS )
+@Target( { ElementType.METHOD, ElementType.ANNOTATION_TYPE } )
+public @interface SubClassMapping {
```

Comment:
Since the term is the single word "[subclass](https://docs.oracle.com/javase/specs/jls/se8/html/jls-8.html#jls-8.1.4)" in Java, wouldn't `SubclassMapping` be a more appropriate camel casing than `SubClassMapping`?

Reply:
I'm fine with it either way, I did not look into how compound words are camelcased in this project.

Here is the reason why I used `SubClassMapping` instead of `SubclassMapping`:
For compound words I look for example if I can put the text ` of a ` in between and retain the meaning.
So for `subclass` I can do this, because `sub of a class` is correct. (in this case sub is short for substitute.)

What do you think, @filiphr ?

Reply:
Although "SubclassMapping" might be "more" correct, I don't like it. The biggest reason why I like `SubClassMapping` more is that in an IDE I would always search by Typing "SB" and "SubclassMapping" is not returned in this case. However, reading more about it and looking at some other projects, it seems "Subclass" is more preferred (e.g. in the ByteBuddy project). 

In any case, lets leave the naming for the end. It would be easy to change it. We'll talk about it with the team.

Reply:
Discussed this in the team and we decided to go with `SubclassMapping`. I can do that change.